packages/rollup/src/index.js

```diff
@@ -0,0 +1,262 @@
+/**
+ * @license
+ * Copyright 2020 The Bazel Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+const path = require('path');
+const rollup = require('rollup');
+const crypto = require('crypto')
+
+const MNEMONIC = 'Rollup';
+const PID = process.pid;
+
+let worker;
+try {
+  worker = require('./worker');
+} catch {
+  // TODO: rely on the linker to link the first-party package
+  const helper = process.env['BAZEL_NODE_RUNFILES_HELPER'];
+  if (!helper) throw new Error('No runfiles helper and no @bazel/worker npm package');
+  const runfiles = require(helper);
+  const workerRequire = runfiles.resolve('build_bazel_rules_nodejs/packages/rollup/worker.js');
+  if (!workerRequire)
+    throw new Error(`build_bazel_rules_nodejs/packages/rollup/worker.js missing in runfiles ${
+        JSON.stringify(runfiles.manifest)}, ${runfiles.dir}`);
+  worker = require(workerRequire);
+}
+
+// Store the cache forever to re-use on each build
+let cacheMap = Object.create(null);
+
+// Generate a unique cache ID based on the input/outputOptions
+function computeCacheKey(inputOptions, outputOptions) {
+  const hash = crypto.createHash('sha256');
+  const hashContent = JSON.stringify(inputOptions) + JSON.stringify(outputOptions);
+  return hash.update(hashContent).digest('hex');
+}
+
+async function runRollup(cacheKeyData, inputOptions, outputOptions) {
+  const cacheKey = computeCacheKey(cacheKeyData);
+
+  let cache = cacheMap[cacheKey];
+
+  const rollupStartTime = Date.now();
+
+  const bundle = await rollup.rollup({...inputOptions, cache});
+
+  const rollupEndTime = Date.now();
+  worker.debug(
+      `${MNEMONIC}[${PID}][${cacheKey}].rollup()`, (rollupEndTime - rollupStartTime) / 1000);
+
+  cacheMap[cacheKey] = bundle.cache;
+
+  try {
+    await bundle.write(outputOptions);
+  } catch (e) {
+    worker.log(e);
+    return false;
+  }
+
+  const bundleEndTime = Date.now();
+  worker.debug(`${MNEMONIC}[${PID}][${cacheKey}].write()`, (bundleEndTime - rollupEndTime) / 1000);
+
+  return true;
+}
+
+// Run rollup, will use + re-populate the cache
+async function runRollupBundler(args /*, inputs */) {
+  const {inputOptions, outputOptions} = await parseCLIArgs(args);
+
+  return runRollup(inputOptions.input, inputOptions, outputOptions);
```

Comment:
The cache key of the main bundle is the input files array. So if other input/outputOptions change it will still share the cache. I think that's ok? I think that's a good thing since those rarely change, and if they do can probably still cache a lot?

Reply:
that sounds scary and incorrect though - if I change the config file my change isn't reflected in the next rebuild? Is that what it means?

Reply:
A cache miss doesn't mean rollup won't run, rollup just won't have any data from a previous run so it must do everything from scratch.

I think the best case would be if the key was the bazel target being executed, so each rollup bazel target has its own cache (actually 2, an extra for the config file).

Although sharing across rules where only the output type changes might be a good thing... meaning what I have here might be best? The same input points should mean the same file dep graph. I think parsing to create that graph would be the expensive thing.

Either way I don't think it would lead to bugs. A cache miss just means rollup must start from scratch. Too many (unrelated?) things sharing a key just means the cache hit might be useless, and then caches something probably useless to the next person as well (like when I had two keys hardcoded).

Reply:
sure, a de-optimization is not so critical as long as the build is correct. I think we still need to iterate on this some more to get the best performance in all cases

Reply:
Yeah I'd like to think about this more. Although I *think* using the bundle-input-files + bundle-config-file (which is also the "input file" for the config) as cache keys for the two rollup invocations might be best. I think the other option would be the specific bazel target.

I'd like to investigate + iterate more on potential perf improvements though 👍